src/net_processing.cpp

```diff
@@ -3844,48 +3857,47 @@ class CompareInvMempoolOrder
 bool PeerLogicValidation::SendMessages(CNode* pto)
 {
     const Consensus::Params& consensusParams = Params().GetConsensus();
-    {
-        // Don't send anything until the version handshake is complete
-        if (!pto->fSuccessfullyConnected || pto->fDisconnect)
-            return true;
 
-        // If we get here, the outgoing message serialization version is set and can't change.
-        const CNetMsgMaker msgMaker(pto->GetSendVersion());
+    if (MaybeDiscourageAndDisconnect(*pto)) return true;
```

Comment:
38c96ef in this commit that removes `MaybeDiscourageAndDisconnect()` from `ProcessMessages()`, future reviewers may appreciate a comment here in the remaining call to make the order dependency explicit.

e.g. that `MaybeDiscourageAndDisconnect()` here depends on all misbehavior calls (currently `Misbehaving`) occurring in ProcessMessages which needs to remain called before SendMessages in `CConnman::ThreadMessageHandler`... if I am not confused.

Reply:
`nMisbehavior` is a tally in `CNodeState` that can be incremented from anywhere. That almost always happens inside a `ProcessMessages()` call (because we increment the misbehavior score when receiving a bad messages from a peer), but not always. See, for example, the call to `MaybePunishNodeForBlock()` inside `BlockChecked()`, which is an asynchronous callback from the validation interface, executed on the scheduler thread.

As long as `MaybeDiscourageAndDisconnect()` is called regularly for the node, then the misbehavior score exceeding the 100 threshold will eventually result in the peer being punished. It doesn't really matter where that `MaybeDiscourageAndDisconnect()` happens, but it makes most sense in `SendMessages()` which is where we do general peer housekeeping/maintenance.

Reply:
Thanks, @jnewbery, makes sense.

The previous behavior before this PR had me wondering if there were particular optimisation reasons for it, but I don't see any.